src/Microsoft.IdentityModel.Protocols.SignedHttpRequest/SignedHttpRequestHandler.cs

```diff
@@ -79,30 +82,78 @@ public string CreateSignedHttpRequest(SignedHttpRequestDescriptor signedHttpRequ
+                }
             }
 
-            // set the "typ" header claim to "pop"
-            // https://datatracker.ietf.org/doc/html/draft-ietf-oauth-signed-http-request-03#section-6.2
-            header[JwtHeaderParameterNames.Alg] = signedHttpRequestDescriptor.SigningCredentials.Algorithm;
-            header[JwtHeaderParameterNames.Typ] = SignedHttpRequestConstants.TokenType;
+            string encodedHeader;
+            using (MemoryStream headerUtf8Bytes = new MemoryStream())
+            {
+                Utf8JsonWriter headerWriter = null;
+                try
+                {
+                    headerWriter = new Utf8JsonWriter(headerUtf8Bytes, new JsonWriterOptions { Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping });
+                    headerWriter.WriteStartObject();
+
+                    if (signedHttpRequestDescriptor.AdditionalHeaderClaims != null && signedHttpRequestDescriptor.AdditionalHeaderClaims.Count != 0)
+                    {
+                        if (signedHttpRequestDescriptor.AdditionalHeaderClaims.Keys.Intersect(JwtTokenUtilities.DefaultHeaderParameters, StringComparer.OrdinalIgnoreCase).Any())
+                            throw LogHelper.LogExceptionMessage(
+                                new SecurityTokenException(
+                                    LogHelper.FormatInvariant(
+                                        JsonWebTokens.LogMessages.IDX14116,
+                                        LogHelper.MarkAsNonPII(nameof(signedHttpRequestDescriptor.AdditionalHeaderClaims)),
+                                        LogHelper.MarkAsNonPII(string.Join(", ", JwtTokenUtilities.DefaultHeaderParameters)))));
+                    }
 
-            var message = Base64UrlEncoder.Encode(Encoding.UTF8.GetBytes(header.ToString(Formatting.None))) + "." + Base64UrlEncoder.Encode(Encoding.UTF8.GetBytes(payload));
+                    // set the "typ" header claim to "pop"
+                    // https://datatracker.ietf.org/doc/html/draft-ietf-oauth-signed-http-request-03#section-6.2
+                    headerWriter.WriteString(JwtHeaderParameterNames.Alg, signedHttpRequestDescriptor.SigningCredentials.Algorithm);
+                    headerWriter.WriteString(JwtHeaderParameterNames.Typ, SignedHttpRequestConstants.TokenType);
+                    if (signedHttpRequestDescriptor.AdditionalHeaderClaims != null)
+                        foreach (string key in signedHttpRequestDescriptor.AdditionalHeaderClaims.Keys)
+                            headerWriter.WriteString(key, signedHttpRequestDescriptor.AdditionalHeaderClaims[key].ToString());
 
-            return message + "." + JwtTokenUtilities.CreateEncodedSignature(message, signedHttpRequestDescriptor.SigningCredentials, false);
+                    headerWriter.WriteEndObject();
+                    headerWriter.Flush();
+                    encodedHeader = Base64UrlEncoder.Encode(headerUtf8Bytes.ToArray());
```

Comment:
one more spot it looks like we can use the streams buffer

Reply:
@keegan-caruso addressed in d789adbb30b8093a0eff70ee545983195935366f